### PR TITLE
planner: introduce a new variable to adjust the penalty double read cost of IndexJoin

### LINKED ISSUE #40556: planner: wrong IndexJoin+IndexReader plan on OSSInsight Workload
## Enhancement
```
EXPLAIN ANALYZE FORMAT='VERBOSE' SELECT
  `r`.`repo_name`,
  count(1) AS `count`
FROM
  `github_events` `ge`
  INNER JOIN `github_repos` `r` ON `ge`.`repo_id` = `r`.`repo_id`
  INNER JOIN `github_users` `u` ON `ge`.`actor_id` = `u`.`id`
WHERE
  `u`.`country_code` = 'CN'
  AND `r`.`primary_language` = 'PHP'
GROUP BY
  `r`.`repo_name`
ORDER BY
  `count` DESC
LIMIT
  20;


+---------------------------------------+--------------+----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| id                                    | estRows      | estCost        | actRows   | task         | access object                                                                                                                                | execution info                                                                                                                                                                                                                                                                                                                                                                                                     | operator info                                                                                                                                                                                                 | memory   | disk    |
+---------------------------------------+--------------+----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| Projection_17                         | 20.00        | 2992846808.96  | 20        | root         |                                                                                                                                              | time:42.3s, loops:2, Concurrency:OFF                                                                                                                                                                                                                                                                                                                                                                               | gharchive_dev.github_repos.repo_name, Column#77                                                                                                                                                               | 27.1 KB  | N/A     |
| └─TopN_20                             | 20.00        | 2992846807.63  | 20        | root         |                                                                                                                                              | time:42.3s, loops:2                                                                                                                                                                                                                                                                                                                                                                                                | Column#77:desc, offset:0, count:20                                                                                                                                                                            | 57.7 KB  | N/A     |
|   └─HashAgg_27                        | 6199650.93   | 1655803580.01  | 130514    | root         |                                                                                                                                              | time:42.3s, loops:136, partial_worker:{wall_time:42.309632445s, concurrency:15, task_num:3750, tot_wait:10m33.179502331s, tot_exec:1.366586188s, tot_time:10m34.602409936s, max:42.309418171s, p95:42.309418171s}, final_worker:{wall_time:42.336074873s, concurrency:15, task_num:225, tot_wait:10m34.600399761s, tot_exec:433.02598ms, tot_time:10m35.03348996s, max:42.335869895s, p95:42.335869895s}           | group by:gharchive_dev.github_repos.repo_name, funcs:count(1)->Column#77, funcs:firstrow(gharchive_dev.github_repos.repo_name)->gharchive_dev.github_repos.repo_name                                          | 105.3 MB | N/A     |
|     └─HashJoin_51                     | 6568709.27   | 1518525536.81  | 3832210   | root         |                                                                                                                                              | time:42.3s, loops:3751, build_hash_table:{total:2.83s, fetch:105.8ms, build:2.72s}, probe:{concurrency:15, total:10m34.5s, max:42.3s, probe:24.7s, fetch:10m9.8s}                                                                                                                                                                                                                                                  | inner join, equal:[eq(gharchive_dev.github_events.repo_id, gharchive_dev.github_repos.repo_id)]                                                                                                               | 737.3 MB | 0 Bytes |
|       ├─TableReader_135(Build)        | 6568709.27   | 251745856.73   | 6663003   | root         |                                                                                                                                              | time:130.4ms, loops:7090, cop_task: {num: 1153, max: 158.3ms, min: 23.9ms, avg: 61.3ms, p95: 103.1ms, rpc_num: 1153, rpc_time: 1m10.6s, copr_cache_hit_ratio: 0.00, distsql_concurrency: 96}                                                                                                                                                                                                                       | data:Selection_134                                                                                                                                                                                            | 43.8 MB  | N/A     |
|       │ └─Selection_134               | 6568709.27   | 22718208663.41 | 6663003   | cop[tiflash] |                                                                                                                                              | tiflash_task:{proc max:134.7ms, min:17.7ms, avg: 54ms, p80:58.6ms, p95:72.7ms, iters:5540, tasks:1153, threads:1153}                                                                                                                                                                                                                                                                                               | eq(gharchive_dev.github_repos.primary_language, "PHP")                                                                                                                                                        | N/A      | N/A     |
|       │   └─TableFullScan_133         | 326019939.00 | 21935760809.81 | 326024222 | cop[tiflash] | table:r                                                                                                                                      | tiflash_task:{proc max:126.7ms, min:17.7ms, avg: 50.1ms, p80:55ms, p95:68.9ms, iters:5541, tasks:1153, threads:1153}, tiflash_scan:{dtfile:{total_scanned_packs:40603, total_skipped_packs:76933, total_scanned_rows:329082557, total_skipped_rows:624158404, total_rs_index_load_time: 0ms, total_read_time: 30922ms}, total_create_snapshot_time: 0ms}                                                           | keep order:false                                                                                                                                                                                              | N/A      | N/A     |
|       └─IndexHashJoin_92(Probe)       | 42448274.98  | 255596779.01   | 156322497 | root         |                                                                                                                                              | time:39.8s, loops:152667, inner:{total:9m19.7s, concurrency:15, task:24, construct:339.3ms, fetch:7m59.1s, build:67.8ms, join:1m20.3s}                                                                                                                                                                                                                                                                             | inner join, inner:IndexReader_89, outer key:gharchive_dev.github_users.id, inner key:gharchive_dev.github_events.actor_id, equal cond:eq(gharchive_dev.github_users.id, gharchive_dev.github_events.actor_id) | 2.27 GB  | N/A     |
|         ├─TableReader_120(Build)      | 392115.00    | 56491346.60    | 412207    | root         |                                                                                                                                              | time:109.3ms, loops:459, cop_task: {num: 138, max: 120.9ms, min: 14.6ms, avg: 47.2ms, p95: 90.7ms, rpc_num: 138, rpc_time: 6.51s, copr_cache_hit_ratio: 0.00, distsql_concurrency: 96}                                                                                                                                                                                                                             | data:Selection_119                                                                                                                                                                                            | 610.3 KB | N/A     |
|         │ └─Selection_119             | 392115.00    | 5384427526.93  | 412207    | cop[tiflash] |                                                                                                                                              | tiflash_task:{proc max:48.4ms, min:9.73ms, avg: 21.2ms, p80:26.8ms, p95:36.6ms, iters:1507, tasks:138, threads:138}                                                                                                                                                                                                                                                                                                | eq(gharchive_dev.github_users.country_code, "CN")                                                                                                                                                             | N/A      | N/A     |
|         │   └─TableFullScan_118       | 95738728.00  | 5154654579.73  | 95738728  | cop[tiflash] | table:u                                                                                                                                      | tiflash_task:{proc max:44.4ms, min:2.08ms, avg: 18.1ms, p80:24.2ms, p95:35.6ms, iters:1507, tasks:138, threads:138}, tiflash_scan:{dtfile:{total_scanned_packs:11757, total_skipped_packs:79, total_scanned_rows:95738728, total_skipped_rows:646901, total_rs_index_load_time: 0ms, total_read_time: 509ms}, total_create_snapshot_time: 0ms}                                                                     | keep order:false                                                                                                                                                                                              | N/A      | N/A     |
|         └─IndexReader_89(Probe)       | 42448274.98  | 459.89         | 156322497 | root         | partition:all                                                                                                                                | time:7m53.3s, loops:159835, cop_task: {num: 55687, max: 5.42s, min: 459.6µs, avg: 632.9ms, p95: 2.16s, max_proc_keys: 50144, p95_proc_keys: 9184, tot_proc: 8h31m14.9s, tot_wait: 1h13m28.2s, rpc_num: 55687, rpc_time: 9h47m21.3s, copr_cache_hit_ratio: 0.01, distsql_concurrency: 96}                                                                                                                           | index:IndexRangeScan_88                                                                                                                                                                                       | 82.8 KB  | N/A     |
|           └─IndexRangeScan_88         | 42448274.98  | 30431.51       | 156322497 | cop[tikv]    | table:ge, index:index_ge_on_actor_id_type_action_created_at_repo_id_commits(actor_id, type, action, created_at, repo_id, push_distinct_size) | tikv_task:{proc max:5.38s, min:0s, avg: 552.6ms, p80:1.14s, p95:2.09s, iters:361540, tasks:55687}, scan_detail: {total_process_keys: 156195201, total_process_keys_size: 25634984161, total_keys: 165830931, get_snapshot_time: 8m22.8s, rocksdb: {delete_skipped_count: 215876, key_skipped_count: 157362521, block: {cache_hit_count: 111735142, read_count: 1106132, read_byte: 11.0 GB, read_time: 10m16.9s}}} | range: decided by [eq(gharchive_dev.github_events.actor_id, gharchive_dev.github_users.id)], keep order:false                                                                                                 | N/A      | N/A     |
+---------------------------------------+--------------+----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
13 rows in set, 2 warnings (42.64 sec)


EXPLAIN ANALYZE FORMAT='VERBOSE' SELECT /*+ read_from_storage(tiflash[ge, u, r]) */
  `r`.`repo_name`,
  count(1) AS `count`
FROM
  `github_events` `ge`
  INNER JOIN `github_repos` `r` ON `ge`.`repo_id` = `r`.`repo_id`
  INNER JOIN `github_users` `u` ON `ge`.`actor_id` = `u`.`id`
WHERE
  `u`.`country_code` = 'CN'
  AND `r`.`primary_language` = 'PHP'
GROUP BY
  `r`.`repo_name`
ORDER BY
  `count` DESC
LIMIT
  20;

+--------------------------------------------------------+---------------+-----------------+------------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+------+
| id                                                     | estRows       | estCost         | actRows    | task         | access object | execution info                                                                                                                                                                                                                                                                                                                                         | operator info                                                                                                                                                        | memory  | disk |
+--------------------------------------------------------+---------------+-----------------+------------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+------+
| Projection_17                                          | 20.00         | 3716065461.65   | 20         | root         |               | time:9.03s, loops:2, Concurrency:OFF                                                                                                                                                                                                                                                                                                                   | gharchive_dev.github_repos.repo_name, Column#77                                                                                                                      | 28.4 KB | N/A  |
| └─TopN_21                                              | 20.00         | 3716065460.32   | 20         | root         |               | time:9.03s, loops:2                                                                                                                                                                                                                                                                                                                                    | Column#77:desc, offset:0, count:20                                                                                                                                   | 1.09 KB | N/A  |
|   └─TableReader_109                                    | 20.00         | 3716060751.04   | 180        | root         | partition:all | time:9.03s, loops:3, cop_task: {num: 18, max: 0s, min: 0s, avg: 0s, p95: 0s, copr_cache_hit_ratio: 0.00}                                                                                                                                                                                                                                               | data:ExchangeSender_108                                                                                                                                              | N/A     | N/A  |
|     └─ExchangeSender_108                               | 20.00         | 356741827743.75 | 180        | mpp[tiflash] |               | tiflash_task:{proc max:9.03s, min:9.03s, avg: 9.03s, p80:9.03s, p95:9.03s, iters:9, tasks:9, threads:144}                                                                                                                                                                                                                                              | ExchangeType: PassThrough                                                                                                                                            | N/A     | N/A  |
|       └─TopN_107                                       | 20.00         | 356741827743.75 | 180        | mpp[tiflash] |               | tiflash_task:{proc max:9.03s, min:9.03s, avg: 9.03s, p80:9.03s, p95:9.03s, iters:9, tasks:9, threads:9}                                                                                                                                                                                                                                                | Column#77:desc, offset:0, count:20                                                                                                                                   | N/A     | N/A  |
|         └─Projection_102                               | 6199650.93    | 356677520975.50 | 130514     | mpp[tiflash] |               | tiflash_task:{proc max:9.03s, min:9.03s, avg: 9.03s, p80:9.03s, p95:9.03s, iters:9, tasks:9, threads:144}                                                                                                                                                                                                                                              | Column#77, gharchive_dev.github_repos.repo_name                                                                                                                      | N/A     | N/A  |
|           └─HashAgg_100                                | 6199650.93    | 356677501136.61 | 130514     | mpp[tiflash] |               | tiflash_task:{proc max:9.03s, min:9.03s, avg: 9.03s, p80:9.03s, p95:9.03s, iters:9, tasks:9, threads:144}                                                                                                                                                                                                                                              | group by:gharchive_dev.github_repos.repo_name, funcs:count(1)->Column#77, funcs:firstrow(gharchive_dev.github_repos.repo_name)->gharchive_dev.github_repos.repo_name | N/A     | N/A  |
|             └─ExchangeReceiver_80                      | 6568709.27    | 356669256807.64 | 3832211    | mpp[tiflash] |               | tiflash_task:{proc max:9.02s, min:9.01s, avg: 9.01s, p80:9.01s, p95:9.02s, iters:538, tasks:9, threads:144}                                                                                                                                                                                                                                            |                                                                                                                                                                      | N/A     | N/A  |
|               └─ExchangeSender_79                      | 6568709.27    | 356071504264.09 | 3832211    | mpp[tiflash] |               | tiflash_task:{proc max:9.15s, min:0s, avg: 1.02s, p80:0s, p95:9.15s, iters:90843, tasks:9, threads:145}                                                                                                                                                                                                                                                | ExchangeType: HashPartition, Hash Cols: [name: gharchive_dev.github_repos.repo_name, collate: utf8mb4_bin]                                                           | N/A     | N/A  |
|                 └─Projection_78                        | 6568709.27    | 356071504264.09 | 3832211    | mpp[tiflash] |               | tiflash_task:{proc max:9.01s, min:0s, avg: 1s, p80:0s, p95:9.01s, iters:90843, tasks:9, threads:145}                                                                                                                                                                                                                                                   | gharchive_dev.github_repos.repo_name                                                                                                                                 | N/A     | N/A  |
|                   └─HashJoin_74                        | 6568709.27    | 356071493754.16 | 3832211    | mpp[tiflash] |               | tiflash_task:{proc max:9.01s, min:0s, avg: 1s, p80:0s, p95:9.01s, iters:90843, tasks:9, threads:145}                                                                                                                                                                                                                                                   | inner join, equal:[eq(gharchive_dev.github_events.repo_id, gharchive_dev.github_repos.repo_id)]                                                                      | N/A     | N/A  |
|                     ├─ExchangeReceiver_51(Build)       | 6568709.27    | 23816234104.94  | 59967027   | mpp[tiflash] |               | tiflash_task:{proc max:2.14s, min:0s, avg: 238.1ms, p80:0s, p95:2.14s, iters:6727, tasks:9, threads:144}                                                                                                                                                                                                                                               |                                                                                                                                                                      | N/A     | N/A  |
|                     │ └─ExchangeSender_50              | 6568709.27    | 22718208663.41  | 6663003    | mpp[tiflash] |               | tiflash_task:{proc max:2.23s, min:0s, avg: 247.8ms, p80:0s, p95:2.23s, iters:5286, tasks:9, threads:144}                                                                                                                                                                                                                                               | ExchangeType: Broadcast                                                                                                                                              | N/A     | N/A  |
|                     │   └─Selection_49                 | 6568709.27    | 22718208663.41  | 6663003    | mpp[tiflash] |               | tiflash_task:{proc max:2.23s, min:0s, avg: 247.8ms, p80:0s, p95:2.23s, iters:5286, tasks:9, threads:144}                                                                                                                                                                                                                                               | eq(gharchive_dev.github_repos.primary_language, "PHP")                                                                                                               | N/A     | N/A  |
|                     │     └─TableFullScan_48           | 326019939.00  | 21935760809.81  | 326024222  | mpp[tiflash] | table:r       | tiflash_task:{proc max:2.19s, min:0s, avg: 243ms, p80:0s, p95:2.19s, iters:5287, tasks:9, threads:144}, tiflash_scan:{dtfile:{total_scanned_packs:40311, total_skipped_packs:14356, total_scanned_rows:326917831, total_skipped_rows:116395683, total_rs_index_load_time: 0ms, total_read_time: 42141ms}, total_create_snapshot_time: 0ms}             | keep order:false                                                                                                                                                     | N/A     | N/A  |
|                     └─HashJoin_42(Probe)               | 42448274.98   | 332170732333.08 | 156322587  | mpp[tiflash] |               | tiflash_task:{proc max:6.43s, min:0s, avg: 714.4ms, p80:0s, p95:6.43s, iters:90843, tasks:9, threads:145}                                                                                                                                                                                                                                              | inner join, equal:[eq(gharchive_dev.github_users.id, gharchive_dev.github_events.actor_id)]                                                                          | N/A     | N/A  |
|                       ├─ExchangeReceiver_46(Build)     | 392115.00     | 5413777334.98   | 3709863    | mpp[tiflash] |               | tiflash_task:{proc max:213.9ms, min:0s, avg: 23.8ms, p80:0s, p95:213.9ms, iters:506, tasks:9, threads:144}                                                                                                                                                                                                                                             |                                                                                                                                                                      | N/A     | N/A  |
|                       │ └─ExchangeSender_45            | 392115.00     | 5384427526.93   | 412207     | mpp[tiflash] |               | tiflash_task:{proc max:218.3ms, min:0s, avg: 24.3ms, p80:0s, p95:218.3ms, iters:1507, tasks:9, threads:112}                                                                                                                                                                                                                                            | ExchangeType: Broadcast                                                                                                                                              | N/A     | N/A  |
|                       │   └─Selection_44               | 392115.00     | 5384427526.93   | 412207     | mpp[tiflash] |               | tiflash_task:{proc max:106.6ms, min:0s, avg: 11.8ms, p80:0s, p95:106.6ms, iters:1507, tasks:9, threads:112}                                                                                                                                                                                                                                            | eq(gharchive_dev.github_users.country_code, "CN")                                                                                                                    | N/A     | N/A  |
|                       │     └─TableFullScan_43         | 95738728.00   | 5154654579.73   | 95738728   | mpp[tiflash] | table:u       | tiflash_task:{proc max:106.3ms, min:0s, avg: 11.8ms, p80:0s, p95:106.3ms, iters:1507, tasks:9, threads:112}, tiflash_scan:{dtfile:{total_scanned_packs:11757, total_skipped_packs:160, total_scanned_rows:95738728, total_skipped_rows:1308879, total_rs_index_load_time: 0ms, total_read_time: 683ms}, total_create_snapshot_time: 0ms}               | keep order:false                                                                                                                                                     | N/A     | N/A  |
|                       └─TableFullScan_47(Probe)        | 5482476251.00 | 317984202558.00 | 5486630011 | mpp[tiflash] | table:ge      | tiflash_task:{proc max:419.9ms, min:0s, avg: 46.7ms, p80:0s, p95:419.9ms, iters:90843, tasks:9, threads:145}, tiflash_scan:{dtfile:{total_scanned_packs:677406, total_skipped_packs:218279, total_scanned_rows:5495771929, total_skipped_rows:1771355390, total_rs_index_load_time: 173ms, total_read_time: 84470ms}, total_create_snapshot_time: 1ms} | keep order:false, PartitionTableScan:true                                                                                                                            | N/A     | N/A  |
+--------------------------------------------------------+---------------+-----------------+------------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+------+
21 rows in set, 3 warnings (9.37 sec)
```

The reason is that the double-read cost caused by `IndexJoin+IndexReader` is not accumulated.
A solution to this issue may be adding penalty costs to cases like this.

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #40556

Problem Summary: planner: introduce a new variable to adjust the penalty double read cost of IndexJoin

### What is changed and how it works?

IndexJoin can cause plenty of extra double-read tasks, which consume lots of resources and take a long time.

Since the number of double-read tasks is hard to be estimated accurately, we introduce a new variable to let us have the ability to adjust this part of the cost manually.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cost_ver2.go [hunk 1]
```diff
@@ -595,7 +596,17 @@ func (p *PhysicalIndexJoin) getIndexJoinCostVer2(taskType property.TaskType, opt
 	batchRatio := 6.0
 	probeCost := divCostVer2(mulCostVer2(probeChildCost, buildRows), batchRatio)
 
-	p.planCostVer2 = sumCostVer2(startCost, buildChildCost, buildFilterCost, buildTaskCost, divCostVer2(sumCostVer2(probeCost, probeFilterCost, hashTableCost), probeConcurrency))
+	// Double Read Cost
+	var doubleReadCost costVer2
+	if p.ctx.GetSessionVars().IndexJoinDoubleReadPenaltyCostRate > 0 {
+		batchSize := float64(p.ctx.GetSessionVars().IndexJoinBatchSize)
+		taskPerBatch := 1024.0 // TODO: remove this magic number
```
Comment (chrysan) [inline planner/core/plan_cost_ver2.go:603]: Why this magic number?

**Review by time-and-fate — APPROVED** [review]

File: planner/core/plan_cost_ver2.go (hunk 1 above)
Comment (qw4990) [reply planner/core/plan_cost_ver2.go:603]: From the OSS case, this magic number can let the planner select a MPP plan when the `penalty-rate=1`.

**Review by chrysan — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Review by fixdb — APPROVED** [review]
+1

**Pull request merged**